orderbook/src/orderbook_pool_fetcher.rs

```diff
@@ -0,0 +1,200 @@
+    }
+}
+
+#[async_trait::async_trait]
+impl PoolFetching for OrderbookPoolFetcher {
+    async fn fetch(&self, pairs: HashSet<TokenPair>, block: BlockNumber) -> Result<Vec<Pool>> {
+        let block = match block {
+            BlockNumber::Number(block) => Some(block.as_u64()),
+            BlockNumber::Latest => None,
+            BlockNumber::Earliest | BlockNumber::Pending => {
+                // Panic because we should never write code that does this.
+                panic!("cache does not support earliest and pending");
+            }
+        };
+
+        let mut cache_hits = Vec::new();
+        let mut cache_misses = HashSet::new();
+        let last_update_block;
+        {
+            let mutexed = self.mutexed.lock().unwrap();
+            for &pair in &pairs {
+                match mutexed.get(pair, block) {
+                    Some(pools) => cache_hits.extend_from_slice(&pools),
+                    None => {
+                        cache_misses.insert(pair);
+                    }
+                }
+            }
+            last_update_block = mutexed.last_update_block;
+        }
+
+        if cache_misses.is_empty() {
+            return Ok(cache_hits);
+        }
+
+        let cache_miss_block = block.unwrap_or(last_update_block);
+        let uncached_pools = self
+            .inner
+            .fetch(
+                cache_misses.clone(),
+                BlockNumber::Number(cache_miss_block.into()),
+            )
+            .await?;
+
+        {
+            let mut mutexed = self.mutexed.lock().unwrap();
+            mutexed.insert(cache_miss_block, pairs.iter().copied(), &uncached_pools);
+            mutexed.mark_used(pairs.iter().copied());
+        }
+
+        cache_hits.extend_from_slice(&uncached_pools);
+        Ok(cache_hits)
+    }
+}
+
+#[async_trait::async_trait]
+impl Maintaining for OrderbookPoolFetcher {
+    async fn run_maintenance(&self, new_block: u64) -> Result<()> {
+        self.update_cache(new_block)
+            .await
+            .context("failed to update pool cache")
+    }
+}
+
+#[derive(Debug)]
+struct Mutexed {
+    // The last N recently used pairs.
+    pairs: LruCache<TokenPair, ()>,
+    // At which block each pair was most recently updated.
+    cached_most_recently_at_block: HashMap<TokenPair, u64>,
+    // Tuple ordering allows us to efficiently construct range queries by block.
+    pools: BTreeMap<(u64, TokenPair), Vec<Pool>>,
+    last_update_block: u64,
+}
+
+impl Mutexed {
+    fn new(pairs_lru_size: usize, current_block: u64) -> Mutexed {
+        Self {
+            pairs: LruCache::new(pairs_lru_size),
+            cached_most_recently_at_block: HashMap::new(),
+            pools: BTreeMap::new(),
+            last_update_block: current_block,
+        }
+    }
+}
+
+impl Mutexed {
+    fn get(&self, pair: TokenPair, block: Option<u64>) -> Option<&[Pool]> {
+        let block = block.or_else(|| self.cached_most_recently_at_block.get(&pair).copied())?;
```

Comment:
Maybe we should more explicitly document this "inexact" matching (I'm slightly afraid that someone will in the future request latest block and assume it is actually going to be the latest block). We could even make the "approximate_matching" a cofigurable boolean on the outer Fetcher struct.

Alternatively we could make BlockNumber in the PoolFetching struct optional, None meaning "I don't care what block"

Reply:
Added a comment. Agree that proper solution is to make this explicit in the block number enum. Would like to do that in future PR.